### PR TITLE
Adding new dialog box changes, courtesy of @time-killer-games.

### PR DESCRIPTION
**Pull request by clee231**
# Description

This pull request includes the changes made by @time-killer-games to incorporate new file dialogs.
Thanks for your contribution! You are awesome!

### CODE REVIEW

File: src/other_libs/dlgmodule/macos/macos.cpp [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+#include <cstdlib>
+#include <cstring>
+#include <cstdio>
+#include <fstream>
+#include <string>
+
+using std::string;
+
+const char *cocoa_widget_get_button_name(int type) {
+  return dialog_module::widget_get_button_name(type);
```
Comment (time-killer-games) [inline src/other_libs/dlgmodule/macos/macos.cpp:41]: i forgot to remove the "dialog_module::" from this line.

File: src/pawgui/pawgui_file_popups.cpp [hunk 2]
```diff
@@ -29,7 +29,12 @@ SOFTWARE.
 -PawByte Ambitious Working GUI(PAWGUI) <https://www.pawbyte.com/pawgui>
 
 
-*/
+*/
+
+#if defined(_WIN32) || (defined(__APPLE__) && defined(__MACH__)) || (defined(__linux__) && !defined(__ANDROID__)) || defined(__FreeBSD__)
+#include "../other_libs/dlgmodule.h"
+#endif
```
Comment (time-killer-games) [inline src/pawgui/pawgui_file_popups.cpp:36]: replace that with:
```
#if defined(_WIN32) || (defined(__APPLE__) && defined(__MACH__)) || (defined(__linux__) && !defined(__ANDROID__)) || defined(__FreeBSD__)
#include <cstring>
#include <SDL2/SDL_syswm.h>
#include "../other_libs/dlgmodule.h"
#include "../sdl_libs/gpe_window_controller_sdl.h"
#endif`
```

File: src/pawgui/pawgui_file_popups.cpp [hunk 3]
```diff
@@ -284,42 +289,16 @@ namespace pawgui
...
+                widget_set_icon((char *)"resources/gamepencil_icon_300dpi.png");
+                std::replace(allowedFileTypes.begin(), allowedFileTypes.end(), '\0', '|');
+                if (!isSaving) {
+                    returnFile = get_open_filename_ext((char *)allowedFileTypes.c_str(), (char *)"", (char *)previousDirectory.c_str(), (char *)prompt.c_str());
+                } else {
+                    returnFile = get_save_filename_ext((char *)allowedFileTypes.c_str(), (char *)"", (char *)previousDirectory.c_str(), (char *)prompt.c_str());
+                }
+                exitOperation = true;
             }
     #endif
```
Comment (time-killer-games) [inline src/pawgui/pawgui_file_popups.cpp:303]: replace that with:
```
    #if defined(_WIN32) || (defined(__APPLE__) && defined(__MACH__)) || (defined(__linux__) && !defined(__ANDROID__)) || defined(__FreeBSD__)
            if (!isDirectorySearch) {
                #if defined(_WIN32) || (defined(__linux__) && !defined(__ANDROID__)) || defined(__FreeBSD__)
                SDL_SysWMinfo wmInfo;
                SDL_VERSION(&wmInfo.version);
                SDL_GetWindowWMInfo(gpe::window_controller_main_sdl->get_sdl_window(), &wmInfo);
                widget_set_icon((char *)"resources/gamepencil_icon_300dpi.png");
                #if defined(_WIN32)
                widget_set_owner((char *)std::to_string((unsigned long long)wmInfo.info.win.window).c_str());
                #elif (defined(__linux__) && !defined(__ANDROID__)) || defined(__FreeBSD__)
                widget_set_owner((char *)std::to_string((unsigned long long)wmInfo.info.x11.window).c_str());
                #endif
                #endif
                char *cFilter = (char *)osFileFilterString.c_str();
                std::string cppFilter; int i = 0;
                while (cFilter[i] != '\0') {
                    cppFilter += &cFilter[i];
                    cppFilter += "|";
                    i += strlen(cFilter + i) + 1;
                }
                if (!cppFilter.empty() && cppFilter.back() == '|')
                    cppFilter.pop_back();
                if (!isSaving) {
                    returnFile = get_open_filename_ext((char *)cppFilter.c_str(), (char *)"", (char *)previousDirectory.c_str(), (char *)prompt.c_str());
                } else {
                    returnFile = get_save_filename_ext((char *)cppFilter.c_str(), (char *)"", (char *)previousDirectory.c_str(), (char *)prompt.c_str());
                }
                exitOperation = true;
            }
    #endif
```

File: src/other_libs/dlgmodule/macos/macos.cpp (hunk 1 above)
Comment (clee231) [reply src/other_libs/dlgmodule/macos/macos.cpp:41]:
```suggestion
  return widget_get_button_name(type);
```
Like this?

Comment (time-killer-games) [reply src/other_libs/dlgmodule/macos/macos.cpp:41]: yes

File: Game_Pencil_Engine_IDE.cbp [hunk 4]
```diff
@@ -0,0 +1,542 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes" ?>
```
Comment (time-killer-games) [inline Game_Pencil_Engine_IDE.cbp:1]: Ok, this time it really is the absolutely last thing. Replace this entire Game_Pencil_Engine.cbp file with:
https://pastebin.com/pAV3uNFQ

File: src/other_libs/dlgmodule/unix/unix.cpp [hunk 5]
```diff
@@ -0,0 +1,1103 @@
+/*
```
Comment (time-killer-games) [inline src/other_libs/dlgmodule/unix/unix.cpp:1]: replace this unix.cpp file with:
https://pastebin.com/sHTy6e4P

File: src/other_libs/dlgmodule/macos/cocoa.mm [hunk 6]
```diff
@@ -0,0 +1,757 @@
+/*
```
Comment (time-killer-games) [inline src/other_libs/dlgmodule/macos/cocoa.mm:1]: delete this file

File: src/other_libs/dlgmodule/macos/macos.cpp [hunk 7]
```diff
@@ -0,0 +1,299 @@
+/*
```
Comment (time-killer-games) [inline src/other_libs/dlgmodule/macos/macos.cpp:1]: delete this file

**Comment by time-killer-games** [comment]
Alright, I have covered absolutely everything. Please add the new commits as soon as you can, and replace the files with the contents I specified as of this post, as I've been updating what I said to replace them with since I initially reviewed.

**Comment by time-killer-games** [comment]
alternately, you can commit that changes made in this tar.xz:
https://drive.google.com/file/d/178PNaxZHz4fjesrfFZjiv1568XQ62QPn/view?usp=sharing

And then I can close https://github.com/pawbyte/Game-Pencil-Engine/pull/40 so that both pr's will be covered in one.

**Comment by time-killer-games** [comment]
Yeah this entire pr can be replaced with this:
https://drive.google.com/file/d/1VB-pIRr9BKZ5TdY7-uJuZh5rVdnW0Dre/view?usp=sharing

I finally fixed the file filter stuff.

**Comment by clee231** [comment]
🤦‍♂️

**Pull request merged**